iocore/net/quic/QUICStreamState.cc

```diff
@@ -211,69 +118,92 @@ QUICReceiveStreamState::update_with_receiving_frame(const QUICFrame &frame)
 }
 
 void
-QUICReceiveStreamState::update(const QUICStreamState &opposite_side)
+QUICReceiveStreamStateMachine::update_on_user_read_event()
 {
   switch (this->get()) {
-  case State::_Init:
-    ink_assert(opposite_side.get() != State::_Init);
-    this->_set_state(State::Recv);
+  case QUICReceiveStreamState::DataRecvd:
+    this->_set_state(QUICReceiveStreamState::DataRead);
     break;
+  case QUICReceiveStreamState::ResetRecvd:
+    this->_set_state(QUICReceiveStreamState::ResetRead);
+    break;
+  case QUICReceiveStreamState::Recv:
+  case QUICReceiveStreamState::SizeKnown:
+  case QUICReceiveStreamState::DataRead:
+  case QUICReceiveStreamState::ResetRead:
+    break;
+  case QUICReceiveStreamState::Init:
   default:
-    ink_assert(!"This shouldn't be happen");
+    ink_assert(!"Unknown state");
     break;
   }
 }
 
 void
-QUICReceiveStreamState::update_on_read()
+QUICReceiveStreamStateMachine::update_on_transport_recv_event()
 {
-  if (this->_in_progress->is_transfer_complete()) {
-    this->_set_state(State::DataRead);
+  if (this->get() == QUICReceiveStreamState::SizeKnown) {
```

Comment:
There is a case of that the recv state jump to Data Recvd (first stream frame has fin flag). Is this covered?
```
 149   case State::_Init:
 150     if (frame.type() == QUICFrameType::STREAM) {
 151       this->_set_state(State::Recv);
 152       if (static_cast<const QUICStreamFrame &>(frame).has_fin_flag()) {
 153         this->_set_state(State::SizeKnown);
 154         if (this->_in_progress->is_transfer_complete()) {
 155           this->_set_state(State::DataRecvd);
 156         }
 157       }
```

Reply:
Yes, it is. Considering this case, Firstly the state is updated in `update_with_receiving_frame` than it transit into SizeKnown (The `Init` check is out of switch blocked, And it immediately transit to Recv). 

If the frame has fin flag, it will transit to SizeKnow immediately. So we only check whether the the state is SizeKnown.

Actually there is a case. That we need to check whether the all data has been read by user in `QUICStream::on_read`.

`update_on_transport_recv_event ` should be call after all data has been received .
`update_on_transport_send_event` should be call after all data has been sent.
`update_on_user_read_event` should be call after all data has been read by user.

Reply:
Good.
Sounds like `update_on_transport_recv_event`/`update_on_transport_send_event` should be `update_on_all_data_recv_event`/`update_on_all_data_sent_event` or something more clear.

Reply:
Like I commented on another thread, you can check "all_data_recv" and "all_data_sent" through `QUICTransferProgressProvide`, so you should not need to introduce new update_on_something if you call existing update_something.